File: src/interpreter/command/assign.ts
import * as cal from 'calcium-js'

export class Assign implements cal.Command {
  constructor(
    public readonly lhs: cal.Assignable,
    public readonly rhs: cal.Expression
  ) {}

  execute(env: cal.Environment): cal.Behavior {
    const rhsValue = env.evaluate(this.rhs)
    this.lhs.assign(rhsValue, env)
    return cal.Behavior.Forward
  }
}

File: lib/interpreter/pseudo/command/assign.dart
import 'package:calcium_app/interpreter/behavior.dart';
import 'package:calcium_app/interpreter/command.dart';
import 'package:calcium_app/interpreter/expression.dart';
import 'package:calcium_app/interpreter/runtime_data.dart';

class Assign implements Command {
  final Reference lhs;
  final Expression rhs;

  Assign({required this.lhs, required this.rhs});

  @override
  Behavior execute(RuntimeData data) {
    lhs.assign(rhs.evaluate(data), data);
    return Behavior.forward;
  }
}
